phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java

```diff
@@ -850,19 +849,12 @@ private void addCoprocessors(byte[] tableName, HTableDescriptor descriptor, PTab
                     && !SchemaUtil.isMetaTable(tableName)
                     && !SchemaUtil.isStatsTable(tableName)) {
                 if (isTransactional) {
-                    if (!descriptor.hasCoprocessor(PhoenixTransactionalIndexer.class.getName())) {
-                        descriptor.addCoprocessor(PhoenixTransactionalIndexer.class.getName(), null, priority, null);
-                    }
```

Comment:
This handles the case of the creation of new tables, but we'd need to handle the case of existing tables as well. This is somewhat trickier because ideally we'd want to handle the case in which the server jar has been upgraded, but the client jar hasn't been. Until the client is upgraded (in which case index maintenance will occur on the client side), we'd need to continue performing index maintenance through this coprocessor. The somewhat clunky way of doing that is to pass the client Phoenix version through the PhoenixIndexMetaData. This would be somewhat painful.

Since we've labelled transactions as "beta", perhaps we can punt on this. In this case, we'd require both the client and server jar to be upgraded at the same time (if transactional tables are already being used). With this approach, we'd want to modify PhoenixTransactionalIndexer to be an empty class derived still from BaseRegionObserver.

Reply:
Maybe I am missing something, but I don't understand why if we upgrade server and client together, we still need an empty PhoenixTransactionalIndexer. Can't we just remove the addCoprocessor from the code and remove PhoenixTransactionalIndexer?

Regarding the case of supporting server upgrade with old clients. Can't we just add a new attribute to each mutation denotes that the index maintenance performed on the client slide and modify PhoenixTransactionalIndexer to ignore mutations (call the super function) for mutations that contains this attribute?

Reply:
The coprocessor is already installed on existing tables. Removing the addCoprocessor only impacts a new table being created. Eventually, we could have some code that runs at upgrade time which removes the coprocessor from existing tables, but we could only do this after we know all clients have been upgraded.

If we want to handle the old client, new server situation, it's slightly more complicated (but not too bad). We have an optimization that conditionally performs an RPC before the batch mutation which contains all the index metadata information (if there are more than a threshold number of mutations being batched). This information is then cached on the RS and looked up by the UUID we store on the Put. That's to prevent having to put information on *every* single mutation. So we'd have to add this attribute to the ServerCache or conditionally add the attribute to the mutation (depending on if we're doing the extra RPC or not).

Reply:
Let's skip this for now and add a release note.